File: examples/java/java-booking-04_tell_dont_ask/src/main/java/tv/codely/booking/Booking.java
package tv.codely.booking;

import java.time.LocalDateTime;

public final class Booking {
    private final BookingId   id;
    private final DateRange   dateRange;
    private final Customer    customer;
    private final BookingType bookingType;
    private final Discount    discount;
    private final Tax         tax;

    public Booking(
        BookingId id,
        DateRange dateRange,
        Customer customer,
        BookingType bookingType,
        Discount discount,
        Tax tax
    ) {
        this.id          = id;
        this.dateRange   = dateRange;
        this.customer    = customer;
        this.bookingType = bookingType;
        this.discount    = discount;
        this.tax         = tax;
    }

    public BookingStatus statusFor(LocalDateTime date) {
        if (date.isBefore(dateRange.startDate())) {
            return BookingStatus.NOT_STARTED;
        }

        if (isBetween(date, dateRange)) {
            return BookingStatus.ACTIVE;
        }

        return BookingStatus.FINISHED;
    }

    private boolean isBetween(LocalDateTime date, DateRange dateRange) {
        return date.isAfter(dateRange.startDate()) && date.isBefore(dateRange.endDate());
    }
}

File: examples/csharp/csharp-booking-03_preserve_whole_object/src/Booking/Booking.cs
using System;

namespace CodelyTv.Booking
{
    public sealed class Booking
    {
        private readonly BookingId id;
        private readonly DateRange dateRange;
        private readonly Customer customer;
        private readonly BookingType bookingType;
        private readonly Discount discount;
        private readonly Tax tax;

        public Booking(
            BookingId id,
            DateRange dateRange,
            Customer customer,
            BookingType bookingType,
            Discount discount,
            Tax tax
        )
        {
            this.id = id;
            this.dateRange = dateRange;
            this.customer = customer;
            this.bookingType = bookingType;
            this.discount = discount;
            this.tax = tax;
        }

        public BookingStatus StatusFor(DateTime date)
        {
            if (date < dateRange.StartDate)
            {
                return BookingStatus.NOT_STARTED;
            }

            if (IsBetween(date, dateRange))
            {
                return BookingStatus.ACTIVE;
            }

            return BookingStatus.FINISHED;
        }

        private bool IsBetween(DateTime date, DateRange dateRange)
        {
            return date > dateRange.StartDate && date < dateRange.EndDate;
        }
    }
}
